ergo-chain-sync/src/cache/redis.rs

```diff
@@ -0,0 +1,200 @@
+use concat_string::concat_string;
+use deadpool_redis::{Config, Pool, Runtime};
+use ergo_lib::{
+    chain::transaction::Transaction,
+    ergo_chain_types::{BlockId, Digest32},
+};
+use redis::{cmd, pipe};
+
+use crate::model::{Block, BlockRecord};
+use async_trait::async_trait;
+
+use super::chain_cache::ChainCache;
+
+static BEST_BLOCK: &str = "best_block";
+
+/// How the blocks are stored in Redis:
+///
+/// Given a block `B`, let `HB` denote the (lowercase) hex-representation of block's ID. Then
+///  - {HB}:p is the key which maps to the hex-representation of B's parent block ID.
+///  - {HB}:h is the key which maps to the height of `B`.
+///  - {HB}:t is the key of a Redis set, which contains the hex-representation `HT` of the
+///    transaction ID of every transaction of `B`.
+///    - Every {HT} is a Redis key which maps to the string-encoded JSON representation of the
+///      transaction.
+///  - {BEST_BLOCK} is a key which maps to a Redis list of length 2, containing the
+///    hex-representations of the best block's ID and its parent ID, in that order.
+pub struct RedisClient {
+    pool: Pool,
+}
+
+impl RedisClient {
+    pub fn new(url: &str) -> Self {
+        let cfg = Config::from_url(url);
+        let pool = cfg.create_pool(Some(Runtime::AsyncStd1)).unwrap();
+        Self { pool }
+    }
+
+    /// Clears out the entire redis DB
+    pub async fn flush_db(&self) {
+        let mut conn = self.pool.get().await.unwrap();
+        let _: () = cmd("FLUSHDB").query_async(&mut conn).await.unwrap();
+    }
+}
+
+#[async_trait(?Send)]
+impl ChainCache for RedisClient {
+    async fn append_block(&mut self, block: Block) {
+        let mut conn = self.pool.get().await.unwrap();
+        let block_id_hex = String::from(block.id.0.clone());
+        let parent_id_hex = String::from(block.parent_id.0.clone());
+        let parent_id_key = concat_string!(block_id_hex, ":p");
+        let block_height_key = concat_string!(block_id_hex, ":h");
+        let block_transactions_key = concat_string!(block_id_hex, ":t");
+
+        let mut pipe = pipe();
+        // SET {block_id_hex}:p {parent_id_hex}
+        pipe.cmd("SET")
+            .arg(&parent_id_key)
+            .arg(&parent_id_hex)
+            .ignore()
+            // SET {block_id_hex}:h {block.height}
+            .cmd("SET")
+            .arg(&block_height_key)
+            .arg(block.height)
+            .ignore();
+
+        for tx in block.transactions {
+            let tx_json = serde_json::to_string(&tx).unwrap();
+            let tx_id = String::from(tx.id().0);
+
+            // SADD {block_id_hex}:t {tx_id}
+            pipe.cmd("SADD")
```

Comment:
A set of commands within a single operation is worth executing in a single transaction, otherwise we will get potentially inconsistent state in the case of unexpected service failure.

Reply:
Sure, it's fixed. Since we can't anticipate something like service failure, should we alter the `ChainCache` trait to return `Result<_, Error>`?